google-cloud-clients/google-cloud-storage/src/main/java/com/google/cloud/storage/BucketInfo.java

```diff
@@ -93,12 +97,302 @@ public BucketInfo apply(com.google.api.services.storage.model.Bucket pb) {
+          return this;
+        }
+
+        /** Sets the number of newer versions a Blob should have for an Action to be executed **/
+        public Builder setNumberOfNewerVersions(Integer numberOfNewerVersions) {
+          this.numberOfNewerVersions = numberOfNewerVersions;
+          return this;
+        }
+
+        /** Sets whether a Live Blob meets the Condition for an Action to be executed. If this is set to {@code false},
+         * the condition is met by Archived Blobs.
+         */
+        public Builder setIsLive(Boolean live) {
+          this.isLive = live;
+          return this;
+        }
+
+        /** Sets a list of Storage Classes ({@code StorageClass} objects) for a Blob to match for an Action to be executed **/
+        public Builder setMatchesStorageClass(List<StorageClass> matchesStorageClass) {
+          this.matchesStorageClass = matchesStorageClass;
+          return this;
+        }
+
+        /** Creates a {@code LifecycleCondition} object **/
+        public LifecycleCondition build() {
+          return new LifecycleCondition(this);
+        }
+      }
+    }
+
+    /**
+     * Base class for the Action to take when a Lifecycle Condition is met. Specific Actions are expressed as subclasses
+     * of this class, accessed by static factory methods
+     */
+    public abstract static class LifecycleAction implements Serializable {
+      private static final long serialVersionUID = 5801228724709173284L;
+
+      public abstract String getActionType();
+
+      /** Creates a new {@code DeleteLifecycleAction}. Blobs that meet the Condition associated with this action will be
+       * deleted
+       */
+      public static DeleteLifecycleAction newDeleteAction() {
+        return new DeleteLifecycleAction();
+      }
+
+      /** Creates a new {@code SetStorageClassLifecycleAction}. Blobs that meet the Condition associated with this
+       * action will have their Storage Class changed to the one specified
+       *
+       * @param storageclass The {@code StorageClass} to set Blobs that meet the condition to
+       */
+      public static SetStorageClassLifecycleAction newSetStorageClassAction(StorageClass storageclass) {
+        return new SetStorageClassLifecycleAction(storageclass);
+      }
+    }
+
+    public static class DeleteLifecycleAction extends LifecycleAction {
+      public static final String TYPE = "Delete";
+      private static final long serialVersionUID = -2050986302222644873L;
+
+      private DeleteLifecycleAction() {}
+
+      @Override
+      public String getActionType() {
+        return TYPE;
+      }
+    }
+
+    public static class SetStorageClassLifecycleAction extends LifecycleAction {
+      public static final String TYPE = "SetStorageClass";
+      private static final long serialVersionUID = -62615467186000899L;
+
+      private final StorageClass storageClass;
+
+      private SetStorageClassLifecycleAction(StorageClass storageClass) {
```

Comment:
Are you okay with changing this name to `SetStorageClassAction` similar with `DeleteAction` given the context?

Reply:
Given that SetStorageClassLifecycleAction is instantiated via LifecycleAction.newSetStorageClassAction(), the long name shouldn't actually have a negative impact so I prefer keeping it as descriptive as possible. But if that's something you strongly disagree with I wouldn't mind changing it.

Reply:
Let's move forward with existing design. Thanks for the feedback!